clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreBufferService.java

```diff
@@ -121,10 +123,27 @@ public int hashCode() {
 
     @Override
     public int getSize() {
-      // TODO:This should not be a big issue but ideally it should calculate the size from leaderProducedRecordContext
-      // if present.
-      return this.consumerRecord.serializedKeySize() + this.consumerRecord.serializedValueSize()
-          + this.consumerRecord.topic().length() + QUEUE_NODE_OVERHEAD_IN_BYTE;
+      // N.B.: This is just an estimate. TODO: Consider if it is really useful, and whether to get rid of it.
+      return this.consumerRecord.getKey().getEstimatedObjectSizeOnHeap()
+          + getEstimateOfMessageEnvelopeSizeOnHeap(this.consumerRecord.getValue()) + QUEUE_NODE_OVERHEAD_IN_BYTE;
+    }
+
+    private int getEstimateOfMessageEnvelopeSizeOnHeap(KafkaMessageEnvelope messageEnvelope) {
+      int kmeBaseOverhead = 100; // Super rough estimate.
+      switch (MessageType.valueOf(messageEnvelope)) {
+        case PUT:
+          Put put = (Put) messageEnvelope.payloadUnion;
+          int size = put.putValue.capacity();
```

Comment:
I saw `PubSubMessage` supports `getPayloadSize`, so should we just use it here?

Reply:
My understanding is that the intent of this code here is to measure the on-heap memory usage, and not only the "useful" part of the memory. That is why I thought it would be more accurate to use `capacity` than `remaining`. That being said, your point is valid that with the `OptimizedKafkaValueSerializer`, then the two BBs are backed by the same array, so `capacity` would be over-counting. I'll add a check to avoid double-counting.